src/main/fc/rc.c

```diff
@@ -489,141 +492,176 @@ static FAST_CODE void processRcSmoothingFilter(void)
-    static bool prevDuplicatePacket[3];             // to identify multiple identical packets
-    static uint16_t feedforwardAveraging = 0;
-
-    if (feedforwardAveraging != pid->feedforwardAveraging) {
-        feedforwardAveraging = pid->feedforwardAveraging;
-        initAveraging(feedforwardAveraging);
-    }
+    static float prevSetpoint[3];                   // equals raw unless extrapolated forward 
+    static float prevSetpointSpeed[3];              // for setpointDelta calculation
+    static float prevSetpointSpeedDelta[3];         // for duplicate extrapolation
+    static bool prevPacketWasADuplicate[3];             // to identify multiple identical packets
 
-    const float rcCommandDeltaAbs = fabsf(rcCommand[axis] - prevRcCommand[axis]);
+    const float rcCommandDelta = rcCommand[axis] - prevRcCommand[axis];
     prevRcCommand[axis] = rcCommand[axis];
+    float rcCommandDeltaAbs = fabsf(rcCommandDelta);
 
-    float setpoint = rawSetpoint[axis];
-    float setpointSpeed = (setpoint - prevSetpoint[axis]);
+    const float setpoint = rawSetpoint[axis];
+    const float setpointDelta = setpoint - prevSetpoint[axis];
     prevSetpoint[axis] = setpoint;
-    float setpointAcceleration = 0.0f;
-
-
-    if (axis == FD_ROLL) {
-        DEBUG_SET(DEBUG_FEEDFORWARD, 0, lrintf(setpoint * 10.0f)); // un-smoothed final feedforward
-    }
-
-    // attenuators
-    float zeroTheAcceleration = 1.0f;
-    float jitterAttenuator = 1.0f;
-    if (pid->feedforwardJitterFactor) {
-        if (rcCommandDeltaAbs < pid->feedforwardJitterFactor) {
-            jitterAttenuator = MAX(1.0f - (rcCommandDeltaAbs + prevRcCommandDeltaAbs[axis]) * pid->feedforwardJitterFactorInv, 0.0f);
-            // note that feedforwardJitterFactorInv includes a divide by 2 to average the two previous rcCommandDeltaAbs values
-            jitterAttenuator = 1.0f - jitterAttenuator * jitterAttenuator;
-        }
-    }
-    prevRcCommandDeltaAbs[axis] = rcCommandDeltaAbs;
 
-    // interpolate setpoint if necessary
-    if (rcCommandDeltaAbs) {
-        // movement!
-        if (prevDuplicatePacket[axis] == true) {
-            rxRate = 1.0f / (rxInterval + prevRxInterval);
-            zeroTheAcceleration = 0.0f;
-            // don't add acceleration, empirically seems better on FrSky
-        }
-        setpointSpeed *= rxRate;
-        prevDuplicatePacket[axis] = false;
-    } else {
-        // no movement!
-        if (prevDuplicatePacket[axis] == false) {
-            // first duplicate after movement
-            setpointSpeed = prevSetpointSpeed[axis];
-            if (fabsf(setpoint) < 0.95f * maxRcRate[axis]) {
-                setpointSpeed += prevAcceleration[axis];
+    float setpointSpeed = 0.0f;
+    float setpointSpeedDelta = 0.0f;
+    float feedforward = 0.0f;
+
+    if (pid->feedforwardInterpolate) {
+        static float prevRxInterval;
+        // for Rx links which do send duplicate data packets when they send telemetry
```

Comment:
It may be better to ignore such packets, as far as rcCommand processing is concerned.

Do you have some trace of such receiver? I'd expect telemetry packets inserted between regular rc packets, with ueven delays, That would break timing assumptions below.

Reply:
This is a comparison of extrapolation 'on' vs 'off' on my FrSky XLite with R-XSR.
Feedforward outcome is quite good.
<img width="1669" alt="FrSky Intrerp on vs off" src="https://github.com/user-attachments/assets/50d00154-73da-44b3-b1fc-ea2797ddf502">
This is the 'gap' we get with telemetry in FrSky... it is exactly 18ms, where normal gap is 9ms.
It's possible that with other protocols we will get irregularity in timing with telemetry, but it seems not a problem for FrSky, at least.

Reply:
So FrSky replaces RC packet with telemetry, missing one value update?

Reply:
It sends what appears to be a normal valid data packet to the FC, but the packet contains the values of the previous packet.
In contrast, ELRS just does 'nothing'; there is a delay that is about twice as long as usual before the new data arrives.